src/transformers/utils/quantizers.py

```diff
@@ -0,0 +1,879 @@
+            # Since weights are saved in the correct "orientation", we skip transposing when loading.
+            if issubclass(module.source_cls, Conv1D):
+                if fp16_statistics is None:
+                    new_value = new_value.T
+
+            kwargs = old_value.__dict__
+            new_value = bnb.nn.Int8Params(new_value, requires_grad=False, **kwargs).to(target_device)
+
+            module._parameters[tensor_name] = new_value
+            if fp16_statistics is not None:
+                setattr(module.weight, "SCB", fp16_statistics.to(target_device))
+
+
+class Bnb4BitHFQuantizer(BnbHFQuantizer):
+    """
+    4-bit quantization from bitsandbytes.py quantization method:
+        before loading: converts transformer layers into Linear4bit during loading: load 16bit weight and pass to the
+        layer object after: quantizes individual weights in Linear4bit into 4bit at fitst .cuda() call
+    serialization: not implemented yet
+    """
+
+    def validate_environment(self, *args, **kwargs):
+        super().validate_environment(*args, **kwargs)
+
+        if version.parse(importlib.metadata.version("bitsandbytes")) < version.parse("0.39.0"):
+            raise ValueError(
+                "You have a version of `bitsandbytes` that is not compatible with 4bit inference and training"
+                " make sure you have the latest version of `bitsandbytes` installed"
+            )
+
+    def adjust_target_dtype(self, target_dtype: torch.dtype) -> torch.dtype:
+        if version.parse(importlib.metadata.version("accelerate")) > version.parse("0.19.0"):
+            from accelerate.utils import CustomDtype
+
+            if target_dtype != torch.int8:
+                logger.info("target_dtype {target_dtype} is replaced by `CustomDtype.INT4` for 4-bit BnB quantization")
+            return CustomDtype.INT4
+        else:
+            raise ValueError(
+                "You are using `device_map='auto'` on a 4bit loaded version of the model. To automatically compute"
+                " the appropriate device map, you should upgrade your `accelerate` library,"
+                "`pip install --upgrade accelerate` or install it from source to support fp4 auto device map"
+                "calculation. You may encounter unexpected behavior, or pass your own device map"
+            )
+
+    def process_model_before_weight_loading(
+        self,
+        model: PreTrainedModel,
+        device_map: DeviceMap,
+        torch_dtype: torch.dtype,
+        keep_in_fp32_modules: List[str] = [],
+    ) -> PreTrainedModel:
+        super().process_model_before_weight_loading(model)
+
+        # TODO: consider moving parts common with 8bits to super()
+        from ..integrations import get_keys_to_not_convert, replace_with_bnb_linear
+
+        llm_int8_skip_modules = (
+            self.quantization_config.llm_int8_skip_modules
+        )  # Despite the name, this still occurs in 4-bit q-configs!
```

Comment:
Yeah, we need to deprecate this arg

Reply:
will remove the comment for now. Will assume that deprecation will happen in some other PR.